src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java

```diff
@@ -0,0 +1,204 @@
+                .toArray(CompoundTopDocs[]::new);
+
+            ScoreNormalizer scoreNormalizer = new ScoreNormalizer();
+            scoreNormalizer.normalizeScores(queryTopDocs, normalizationTechnique);
+
+            ScoreCombiner scoreCombinator = new ScoreCombiner();
+            scoreCombinator.combineScores(topDocsAndMaxScores, queryTopDocs, combinationTechnique);
+
+            updateOriginalQueryResults(searchPhaseResult, queryTopDocs);
+        }
+    }
+
+    @Override
+    public SearchPhaseName getBeforePhase() {
+        return SearchPhaseName.QUERY;
+    }
+
+    @Override
+    public SearchPhaseName getAfterPhase() {
+        return SearchPhaseName.FETCH;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public String getTag() {
+        return tag;
+    }
+
+    @Override
+    public String getDescription() {
+        return description;
+    }
+
+    @Override
+    public boolean isIgnoreFailure() {
+        return true;
+    }
+
+    protected void validateParameters(final String normalizationTechniqueName, final String combinationTechniqueName) {
+        if (Strings.isNullOrEmpty(normalizationTechniqueName)) {
+            throw new IllegalArgumentException("normalization technique cannot be empty");
+        }
+        if (Strings.isNullOrEmpty(combinationTechniqueName)) {
+            throw new IllegalArgumentException("combination technique cannot be empty");
+        }
+        if (!EnumUtils.isValidEnum(ScoreNormalizationTechnique.class, normalizationTechniqueName)) {
+            log.error(String.format(Locale.ROOT, "provided normalization technique [%s] is not supported", normalizationTechniqueName));
+            throw new IllegalArgumentException("provided normalization technique is not supported");
+        }
+        if (!EnumUtils.isValidEnum(ScoreCombinationTechnique.class, combinationTechniqueName)) {
+            log.error(String.format(Locale.ROOT, "provided combination technique [%s] is not supported", combinationTechniqueName));
+            throw new IllegalArgumentException("provided combination technique is not supported");
+        }
+    }
+
+    private boolean isNotHybridQuery(final Optional<SearchPhaseResult> maybeResult) {
+        return maybeResult.isEmpty()
+            || Objects.isNull(maybeResult.get().queryResult())
+            || !(maybeResult.get().queryResult().topDocs().topDocs instanceof CompoundTopDocs);
+    }
+
+    private <Result extends SearchPhaseResult> TopDocsAndMaxScore[] getCompoundQueryTopDocsForResult(
+        final SearchPhaseResults<Result> results
+    ) {
+        List<Result> preShardResultList = results.getAtomicArray().asList();
+        TopDocsAndMaxScore[] result = new TopDocsAndMaxScore[results.getAtomicArray().length()];
+        int idx = 0;
+        for (Result shardResult : preShardResultList) {
+            if (shardResult == null) {
+                idx++;
```

Comment:
one thing I am not understanding here that if results are null we are moving the idx via idx++, same at line 178 and 179, why we are doing that?

Reply:
it's more like temporary int index for iterating over collection in loop. I'll rewrite it to explicit for loop to make it more clear